### PR TITLE
refactor to create value only if it wasn't set

### PR DESCRIPTION
**Pull request by MaximBazarov**

### CODE REVIEW

File: Sources/Recoil/Family.swift [hunk 1]
```diff
@@ -2,21 +2,32 @@ import Foundation
...
+
+    public init(newID: @escaping () -> ID, newAtom: @escaping () -> Atom<ID, T>) {
+        self.newID = newID
+        self.newAtom = newAtom
     }
     
-    public func append(id: ID, value: () -> T) {
-        let atom = Atom { value() }
-        items[id] = atom
+    public func append(value: T) -> ID {
```
Comment (AlexeyDemedetskiy) [inline Sources/Recoil/Family.swift:14]: We do not need append anymore

File: Sources/Recoil/Family.swift [hunk 2]
```diff
@@ -2,21 +2,32 @@ import Foundation
...
-            return nil
+
+    /// Returns the atom for provided id **or creates a new one and returns
+    /// - Parameter id: id to obtain the atom
+    /// - Returns: Atom stored for the id or a new atom
+    ///
+    public func at(id: ID) -> Atom<ID, T> {
+        guard let atom = atoms[id] else {
+            let atom = newAtom()
+            let id = newID()
```
Comment (AlexeyDemedetskiy) [inline Sources/Recoil/Family.swift:27]: no need for newId, you already have id in params

File: Sources/Recoil/Family.swift [hunk 3]
```diff
@@ -2,21 +2,32 @@ import Foundation
...
 public class Family<ID: Hashable, T: Codable> {
 
-    private var items = [ID: Atom<T>]()
-    
-    public init() {
+    private var atoms = [ID: Atom<ID, T>]()
+    private var newID: () -> ID
+    private var newAtom: () -> Atom<ID, T>
+
+    public init(newID: @escaping () -> ID, newAtom: @escaping () -> Atom<ID, T>) {
```
Comment (AlexeyDemedetskiy) [inline Sources/Recoil/Family.swift:9]: Atom should be constructed inside family

File: Sources/Recoil/Atom.swift [hunk 4]
```diff
@@ -2,13 +2,15 @@ import Combine
 import Foundation
 import SwiftUI
 
-public class Atom<T: Codable>: ObservableObject {
+public class Atom<ID: Hashable, T: Codable>: ObservableObject {
```
Comment (AlexeyDemedetskiy) [inline Sources/Recoil/Atom.swift:5]: Atom do not need any id

**Review by AlexeyDemedetskiy — CHANGES_REQUESTED** [review]

File: Sources/Recoil/Atom.swift [hunk 5]
```diff
@@ -4,11 +4,10 @@ import SwiftUI
 
 public class Atom<T: Codable>: ObservableObject {
 
-    public let id = UUID()
-    
-    public init(initial: () -> T) {
-        self.value = initial()
+    @Published var value: T
+
+    public init(value: T) {
```
Comment (AlexeyDemedetskiy) [inline Sources/Recoil/Atom.swift:9]: Should be closure

File: Sources/Recoil/Family.swift [hunk 6]
```diff
@@ -2,23 +2,26 @@ import Foundation
 
 public class Family<ID: Hashable, T: Codable> {
 
-    private var items = [ID: Atom<T>]()
-    
-    public init() {
+    public init(initial: @escaping () -> Atom<T>) {
```
Comment (AlexeyDemedetskiy) [inline Sources/Recoil/Family.swift:5]: Should be `(id) -> T` not `() -> Atom<T>`

**Review by AlexeyDemedetskiy — CHANGES_REQUESTED** [review]

File: Sources/Recoil/Atom.swift (hunk 5 above)
Comment (MaximBazarov) [reply Sources/Recoil/Atom.swift:9]: why?

File: Sources/Recoil/Family.swift [hunk 7]
```diff
@@ -2,23 +2,26 @@ import Foundation
 
 public class Family<ID: Hashable, T: Codable> {
 
-    private var items = [ID: Atom<T>]()
-    
-    public init() {
+    public init(initial: @escaping () -> T) {
```
Comment (AlexeyDemedetskiy) [inline Sources/Recoil/Family.swift:5]: should receive id

**Review by AlexeyDemedetskiy — CHANGES_REQUESTED** [review]

**Review by AlexeyDemedetskiy — APPROVED** [review]

**Pull request merged**